streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1858,10 +1860,15 @@ object ZStream extends ZStreamPlatformSpecific {
    */
   final def managed[R, E, A](managed: ZManaged[R, E, A]): ZStream[R, E, A] =
     new ZStream[R, E, A] {
-      def fold[R1 <: R, E1 >: E, A1 >: A, S]: Fold[R1, E1, A1, S] =
-        ZManaged.succeed { (s, cont, f) =>
-          if (!cont(s)) ZManaged.succeed(s)
-          else managed.mapM(f(s, _))
+      def fold[R1 <: R, E1 >: E, A1 >: A, S]: Fold[R1, E1, A1, S] = foldDefault
+
+      override def process: ZManaged[R, E, InputStream[E, A]] =
+        for {
+          p <- Promise.make[Nothing, Unit].toManaged_
```

Comment:
Why not just `Ref[Boolean]` that represents if the `InputStream` has been pulled or not?

Reply:
Sure, no preference, just wanted to have some fun.

Reply:
Ah, ok, it's just much less efficient :-)

Reply:
Great, changing it.

Reply:
Done.